File: typescript/tests/prekey_persistence.test.ts
/**
 * Persistence of identity keys, signed-pre-key history, and the OPK pool
 * via the {@link PreKeyStore} adapter — verifies that an instance which
 * restarts against the same KV store recovers all responder-side X3DH
 * material so peers can still mint new pre-key bundles AND complete
 * already-issued ones.
 *
 * Mirrors the C# unit-suite {@code SignalProtocolService.PreKeyPersistenceTests}.
 *
 * SPDX-License-Identifier: MIT
 */
import { describe, it } from "node:test";
import { strict as assert } from "node:assert";

import { SignalProtocol } from "../src/security/SignalProtocol.js";
import {
  InMemorySignalSessionStore,
  KeyValueSignalSessionStore,
} from "../src/security/SignalSessionStore.js";
import {
  InMemoryPreKeyStore,
  KeyValuePreKeyStore,
} from "../src/security/PreKeyStore.js";
import { InMemoryKeyValueStore } from "../src/storage/index.js";

describe("PreKeyStore — identity persistence", () => {
  it("first boot persists the freshly-generated identity", async () => {
    const prekeys = new InMemoryPreKeyStore();
    const svc = new SignalProtocol({ preKeyStore: prekeys });
    await svc.ready();
    const stored = await prekeys.loadIdentity();
    assert.ok(stored !== null);
    assert.equal(stored!.ed25519PublicKey.length, 32);
    assert.equal(stored!.x25519PublicKey.length, 32);
  });

  it("a second instance loads the same identity rather than generating a new one", async () => {
    const prekeys = new InMemoryPreKeyStore();
    const a = new SignalProtocol({ preKeyStore: prekeys });
    await a.ready();
    await a.flushPendingWrites();
    const aPub = Buffer.from(a.getPublicKey()).toString("hex");
    const aXPub = Buffer.from(a.getX25519PublicKey()).toString("hex");

    const b = new SignalProtocol({ preKeyStore: prekeys });
    await b.ready();
    assert.equal(Buffer.from(b.getPublicKey()).toString("hex"), aPub);
    assert.equal(Buffer.from(b.getX25519PublicKey()).toString("hex"), aXPub);
  });
});

describe("PreKeyStore — SPK history persistence", () => {
  it("restored instance reuses the active SPK rather than rotating immediately", async () => {
    const prekeys = new InMemoryPreKeyStore();
    let svc = new SignalProtocol({ preKeyStore: prekeys });
    const b1 = await svc.generatePreKeyBundle("alice");
    await svc.flushPendingWrites();

    svc = new SignalProtocol({ preKeyStore: prekeys });
    const b2 = await svc.generatePreKeyBundle("alice");
    assert.equal(b1.signedPreKeyId, b2.signedPreKeyId);
    assert.deepEqual(Array.from(b1.signedPreKey), Array.from(b2.signedPreKey));
  });

  it("retained-history entry survives restart and still decrypts late messages", async () => {
    // Force rotation between bundle issuance and message arrival —
    // proves the responder can still complete X3DH against an SPK that
    // is no longer the active one but is within the retained window.
    const prekeys = new InMemoryPreKeyStore();
    const sessions = new InMemorySignalSessionStore();
    let now = new Date(2026, 0, 1).getTime();
    const nowProvider = () => new Date(now);

    const bobBuild = () =>
      new SignalProtocol({
        preKeyStore: prekeys,
        sessionStore: sessions,
        rotationOptions: { rotationIntervalMs: 1000, retainedHistoryCount: 3 },
        nowProvider,
      });

    let bob = bobBuild();
    const oldBundle = await bob.generatePreKeyBundle("bob");
    const alice = new SignalProtocol();
    await alice.generatePreKeyBundle("alice");
    await alice.processPreKeyBundle(oldBundle);
    const ct = await alice.encrypt("bob", new Uint8Array(Buffer.from("late arrival")));

    // Time advances past the rotation interval; Bob rotates on the next
    // bundle call, then is restarted before the late message arrives.
    now += 5000;
    await bob.generatePreKeyBundle("bob");
    await bob.flushPendingWrites();

    bob = bobBuild();
    await bob.ready();
    // History size should be 2 (active + 1 retained).
    assert.equal(bob.signedPreKeyHistoryCount, 2);

    const pt = await bob.decrypt("alice", ct);
    assert.deepEqual(Array.from(pt), Array.from(Buffer.from("late arrival")));
  });

  it("KV-backed SPK history serialises to and from JSON faithfully", async () => {
    const kv = new InMemoryKeyValueStore();
    const prekeys = new KeyValuePreKeyStore(kv);
    let svc = new SignalProtocol({ preKeyStore: prekeys });
    const b1 = await svc.generatePreKeyBundle("alice");
    await svc.flushPendingWrites();

    // Inspect the stored JSON for sanity.
    const raw = await kv.get(KeyValuePreKeyStore.SPK_HISTORY_KEY);
    assert.ok(raw !== null);
    const parsed = JSON.parse(Buffer.from(raw!).toString("utf8"));
    assert.equal(parsed.entries.length, 1);
    assert.equal(parsed.entries[0].id, b1.signedPreKeyId);

    // Restart and verify reuse.
    svc = new SignalProtocol({ preKeyStore: prekeys });
    const b2 = await svc.generatePreKeyBundle("alice");
    assert.equal(b1.signedPreKeyId, b2.signedPreKeyId);
  });
});

describe("PreKeyStore — OPK pool persistence", () => {
  it("issued-but-unconsumed OPKs are restored in the issued state after restart", async () => {
    const prekeys = new InMemoryPreKeyStore();
    let bob = new SignalProtocol({ preKeyStore: prekeys, opkPoolSize: 5 });
    const bundle = await bob.generatePreKeyBundle("bob");
    await bob.flushPendingWrites();
    // Pool layout: 4 available + 1 issued = 5 held.
    assert.equal(bob.getOpkPoolStatus().held, 5);
    assert.equal(bob.getOpkPoolStatus().available, 4);

    bob = new SignalProtocol({ preKeyStore: prekeys, opkPoolSize: 5 });
    await bob.ready();
    assert.equal(bob.getOpkPoolStatus().held, 5);
    assert.equal(bob.getOpkPoolStatus().available, 4);

    // Bob can still complete X3DH against the issued OPK.
    const alice = new SignalProtocol();
    await alice.generatePreKeyBundle("alice");
    await alice.processPreKeyBundle(bundle);
    const m1 = await alice.encrypt("bob", new Uint8Array(Buffer.from("hi")));
    const p1 = await bob.decrypt("alice", m1);
    assert.deepEqual(Array.from(p1), Array.from(Buffer.from("hi")));
  });

  it("OPKs consumed before restart are NOT held after restart", async () => {
    const prekeys = new InMemoryPreKeyStore();
    let bob = new SignalProtocol({
      preKeyStore: prekeys,
      sessionStore: new InMemorySignalSessionStore(),
      opkPoolSize: 5,
    });
    const bundle = await bob.generatePreKeyBundle("bob");

    const alice = new SignalProtocol();
    await alice.generatePreKeyBundle("alice");
    await alice.processPreKeyBundle(bundle);
    const m1 = await alice.encrypt("bob", new Uint8Array([1]));
    await bob.decrypt("alice", m1);
    await bob.flushPendingWrites();

    // Held drops to 4 after consumption.
    assert.equal(bob.getOpkPoolStatus().held, 4);

    bob = new SignalProtocol({
      preKeyStore: prekeys,
      sessionStore: new InMemorySignalSessionStore(),
      opkPoolSize: 5,
    });
    await bob.ready();
    assert.equal(bob.getOpkPoolStatus().held, 4);
    assert.equal(bob.getOpkPoolStatus().available, 4);
  });
});

File: go/security/prekey_persistence_test.go
// SPDX-License-Identifier: MIT

package security

import (
	"bytes"
	"context"
	"testing"

	"github.com/bhengubv/aether-protocol/go/storage"
)

// TestPreKeyPersistence_IdentityAndSPKAndOPKsSurviveRestart establishes
// pre-key state, simulates a restart, and verifies the new service has
// the same identity, the same active SPK id, and the same OPK pool.
func TestPreKeyPersistence_IdentityAndSPKAndOPKsSurviveRestart(t *testing.T) {
	ctx := context.Background()
	kv := storage.NewInMemoryKeyValueStore()
	store := NewKVPreKeyStore(kv)

	// Phase 1.
	bob1, err := NewSignalProtocolService(WithPreKeyStore(store))
	if err != nil {
		t.Fatalf("bob1: %v", err)
	}
	bundle1, err := bob1.GeneratePreKeyBundle("bob")
	if err != nil {
		t.Fatalf("GeneratePreKeyBundle: %v", err)
	}

	identity1, err := store.LoadIdentity(ctx)
	if err != nil {
		t.Fatalf("LoadIdentity: %v", err)
	}
	if identity1 == nil {
		t.Fatalf("LoadIdentity after first bundle: got nil")
	}
	if identity1.LocalUhid != "bob" {
		t.Errorf("LocalUhid: got %q want bob", identity1.LocalUhid)
	}

	history1, err := store.LoadSignedPreKeys(ctx)
	if err != nil {
		t.Fatalf("LoadSignedPreKeys: %v", err)
	}
	if len(history1.Entries) != 1 {
		t.Fatalf("history entries: got %d want 1", len(history1.Entries))
	}
	if history1.Entries[0].ID != bundle1.SignedPreKeyID {
		t.Errorf("history[0].ID: got %d want %d", history1.Entries[0].ID, bundle1.SignedPreKeyID)
	}

	opks1, err := store.LoadOneTimePreKeys(ctx)
	if err != nil {
		t.Fatalf("LoadOneTimePreKeys: %v", err)
	}
	if len(opks1) != DefaultOpkPoolSize {
		t.Errorf("opks: got %d want %d", len(opks1), DefaultOpkPoolSize)
	}

	// Phase 2: restart.
	bob2, err := NewSignalProtocolService(WithPreKeyStore(store))
	if err != nil {
		t.Fatalf("bob2: %v", err)
	}

	// Identity must match exactly.
	if !bytes.Equal(bob1.GetX25519PublicKey(), bob2.GetX25519PublicKey()) {
		t.Errorf("X25519 identity differs across restart")
	}
	if !bytes.Equal(bob1.GetPublicKey(), bob2.GetPublicKey()) {
		t.Errorf("Ed25519 identity differs across restart")
	}

	// Active SPK id must match.
	if bob2.ActiveSignedPreKeyID() != bundle1.SignedPreKeyID {
		t.Errorf("ActiveSignedPreKeyID: got %d want %d",
			bob2.ActiveSignedPreKeyID(), bundle1.SignedPreKeyID)
	}

	// OPK pool size must match (ignoring the one OPK we already issued).
	held, available := bob2.GetOpkPoolStatus()
	if held != DefaultOpkPoolSize {
		t.Errorf("held after restart: got %d want %d", held, DefaultOpkPoolSize)
	}
	// One OPK was issued in phase 1 but not yet consumed — it stays in held but
	// is not in available.
	if available != DefaultOpkPoolSize-1 {
		t.Errorf("available after restart: got %d want %d", available, DefaultOpkPoolSize-1)
	}
}

// TestPreKeyPersistence_OpkConsumptionPersists verifies that after a peer
// consumes an OPK via X3DH, restarting the responder reflects the
// consumption — the consumed OPK is gone.
func TestPreKeyPersistence_OpkConsumptionPersists(t *testing.T) {
	ctx := context.Background()
	kv := storage.NewInMemoryKeyValueStore()
	store := NewKVPreKeyStore(kv)
	sessionKV := storage.NewInMemoryKeyValueStore()
	sessions := NewKVSessionStore(sessionKV)

	bob1, err := NewSignalProtocolService(WithPreKeyStore(store), WithSessionStore(sessions))
	if err != nil {
		t.Fatalf("bob1: %v", err)
	}
	bundle, err := bob1.GeneratePreKeyBundle("bob")
	if err != nil {
		t.Fatalf("GeneratePreKeyBundle: %v", err)
	}
	consumedOpkID := bundle.PreKeyID

	alice, _ := NewSignalProtocolService()
	alice.GeneratePreKeyBundle("alice")
	if err := alice.ProcessPreKeyBundle(bundle); err != nil {
		t.Fatalf("ProcessPreKeyBundle: %v", err)
	}

	// First message — initiator-side X3DH; bob1 consumes the OPK on decrypt.
	first, _ := alice.Encrypt("bob", []byte("first"))
	if _, err := bob1.Decrypt("alice", first); err != nil {
		t.Fatalf("bob1.Decrypt: %v", err)
	}

	// Verify in-memory and on-disk: consumed OPK gone.
	if _, ok := bob1.preKeys.oneTimePreKeys[consumedOpkID]; ok {
		t.Errorf("consumed OPK %d still in bob1 in-memory pool", consumedOpkID)
	}
	opks, err := store.LoadOneTimePreKeys(ctx)
	if err != nil {
		t.Fatalf("LoadOneTimePreKeys: %v", err)
	}
	if _, present := opks[consumedOpkID]; present {
		t.Errorf("consumed OPK %d still in persisted pool", consumedOpkID)
	}

	// Restart — consumed OPK must stay gone.
	bob2, err := NewSignalProtocolService(WithPreKeyStore(store), WithSessionStore(sessions))
	if err != nil {
		t.Fatalf("bob2: %v", err)
	}
	if _, ok := bob2.preKeys.oneTimePreKeys[consumedOpkID]; ok {
		t.Errorf("consumed OPK %d resurrected after restart", consumedOpkID)
	}
}

// TestPreKeyPersistence_SetLocalUhidPersists verifies the UHID is saved
// alongside identity keys when explicitly set.
func TestPreKeyPersistence_SetLocalUhidPersists(t *testing.T) {
	ctx := context.Background()
	kv := storage.NewInMemoryKeyValueStore()
	store := NewKVPreKeyStore(kv)

	svc, err := NewSignalProtocolService(WithPreKeyStore(store))
	if err != nil {
		t.Fatalf("svc: %v", err)
	}
	svc.SetLocalUhid("explicit-uhid")

	stored, err := store.LoadIdentity(ctx)
	if err != nil {
		t.Fatalf("LoadIdentity: %v", err)
	}
	if stored == nil {
		t.Fatalf("LoadIdentity: got nil")
	}
	if stored.LocalUhid != "explicit-uhid" {
		t.Errorf("LocalUhid: got %q want explicit-uhid", stored.LocalUhid)
	}
}
